### PR TITLE
Fix Issue 18439 Fix issue with `static foreach` not working with scoped @nogc.

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
This should go in stable.

**Comment by tgehr** [comment]
Yuck. This is the wrong fix.

1. The issue is not with `static foreach`, it's a general CTFE problem.
2. This only fixes it for @nogc. There are other attributes.

**Comment by look-at-me** [comment]
> The issue is not with static foreach, it's a general CTFE problem.

Which is not easily fixed, `@nogc` is a blunt instrument. In this case the `@nogc ` is caused because of the hidden code that the compiler generates. Which the user has no access to and generates a very obsure error message. Sure it is not pretty, but this fix is better than waiting year(s?) for an actual fix.
 
> This only fixes it for @nogc. There are other attributes.

None of the other attributes cause `static foreach` to fail.

File: src/dmd/cond.d [hunk 1]
```diff
@@ -288,8 +288,18 @@ extern (C++) final class StaticForeach : RootObject
...
-    private void lowerNonArrayAggregate(Scope* sc)
+    private void lowerNonArrayAggregate(Scope* scInput)
     {
+        Scope* sc = scInput;
+
+        if (sc.stc & STC.nogc)
+        {
+            // concating `~=` an array is only done at compile time for `static foreach` range
+            // remove @nogc to avoid a compile error in this case
+            sc = scInput.copy();
```
Comment (WalterBright) [inline src/dmd/cond.d]: should use `push`, not `copy`

**Comment by tgehr** [comment]
> > The issue is not with static foreach, it's a general CTFE problem.
> 
> Which is not easily fixed,

I doubt it, but I am too tired to investigate right now. @safe does not have the problem. Use the same solution for all other attributes.

> `@nogc` is a blunt instrument. In this case the `@nogc` error is caused because of the hidden code that the compiler generates. Which the user has no access to and generates a very obsure error message. Sure it is not pretty, but this fix is better than waiting year(s?) for an actual fix.

This would be the case if the code had a comment specifying that this is a hack that should be reverted asap.

> 
> > This only fixes it for @nogc. There are other attributes.
> 
> None of the other attributes cause `static foreach` to fail.

```
struct Iota{
	int s,e;
	void popFront(){ s++; }
	@property int front(){ return s; }
	@property bool empty(){ return s>=e; }
}
pure:
void main(){
	static foreach(i;Iota(0,1)){}
}
```

**Comment by look-at-me** [comment]
> I doubt it, but I am too tired to investigate right now. @safe does not have the problem. Use the same solution for all other attributes.

`@safe` can't be auto inferred by templates.

> This would be the case if the code had a comment specifying that this is a hack that should be reverted asap.

I don't think it is necessary, keep me posted if you do end up resolving the general CTFE issue though.

> ```D
> struct Iota{
> 	int s,e;
> 	void popFront(){ s++; }
> 	@property int front(){ return s; }
> 	@property bool empty(){ return s>=e; }
> }
> pure:
> void main(){
> 	static foreach(i;Iota(0,1)){}
> }
> ```

This is in part an issue with the user code as it is not `pure`. It can be worked around by making the methods pure. You can't work around the issue with `@nogc` because the code that is using the GC was the hidden code the compiler generates.

**Comment by tgehr** [comment]
> > This would be the case if the code had a comment specifying that this is a hack that should be reverted asap.
> 
> I don't think it is necessary, keep me posted if you do end up resolving the general CTFE issue though.
> ...

The following change will do the job for @nogc, but not `pure`. It seems the logic for inference of `pure` has diverged from the inference logic of the other attributes, which is horrible.

```
diff --git a/src/dmd/semantic3.d b/src/dmd/semantic3.d
index 764a4a158..1ae8880de 100644
--- a/src/dmd/semantic3.d
+++ b/src/dmd/semantic3.d
@@ -315,8 +315,7 @@ private extern(C++) final class Semantic3Visitor : Visitor
             sc2.fes = funcdecl.fes;
             sc2.linkage = LINK.d;
             sc2.stc &= ~(STC.auto_ | STC.scope_ | STC.static_ | STC.extern_ | STC.abstract_ | STC.deprecated_ | STC.override_ |
-                         STC.TYPECTOR | STC.final_ | STC.tls | STC.gshared | STC.ref_ | STC.return_ | STC.property |
-                         STC.nothrow_ | STC.pure_ | STC.safe | STC.trusted | STC.system);
+                         STC.TYPECTOR | STC.FUNCATTR | STC.final_ | STC.tls | STC.gshared | STC.return_);
             sc2.protection = Prot(Prot.Kind.public_);
             sc2.explicitProtection = 0;
             sc2.aligndecl = null;
```

> > ```d
> > struct Iota{
> > 	int s,e;
> > 	void popFront(){ s++; }
> > 	@property int front(){ return s; }
> > 	@property bool empty(){ return s>=e; }
> > }
> > pure:
> > void main(){
> > 	static foreach(i;Iota(0,1)){}
> > }
> > ```
> 
> This is in part an issue with the user code as it is not `pure`.

No, there is *no issue* with the user code. The range can be traversed in CTFE without any problem.

> It can be worked around by making the methods pure.

The methods may not be in your own code.

> You can't work around the issue with `@nogc` because the code that is using the GC was the hidden code the compiler generates.

Yes, it can be worked around, by manually computing the array.

**Comment by look-at-me** [comment]
@tgehr Any update on your fix?

**Comment by tgehr** [comment]
@look-at-me Very busy until Thursday. The above patch properly fixes the issue for `@nogc` (like you were going for). You can create a pull request with this if you want, otherwise I'll take it eventually, probably on Friday.

**Comment by look-at-me** [comment]
@tgehr 
Your patch has different behavior than what DMD currently does. The `@nogc` attribute isn't applied to the nested lambda like it previous did.

```D
@nogc:
auto test() {
    return () {
        // should be @nogc and error but doesn't
        new int;
    };
}
```

**Comment by tgehr** [comment]
@look-at-me If you are fine with DMD' current behavior, we don't need a patch and the bug report should be closed as invalid. Your `@nogc`  test function is not actually allocating anything, so there shouldn't be an error there. Recursively applying attributes to nested functions is a stupid idea as their attributes are inferred anyway.

**Comment by look-at-me** [comment]
> If we are fine with DMD's current behavior, we don't need a patch and the bug report should be closed as invalid. 

The feature shouldn't be using a lambda in the first place. The reason it does is because the person implementing it took a shortcut. 

> Recursively applying attributes to nested functions is a stupid idea as their attributes are inferred anyway.

I agree, but it's a big-ish breaking change. I can already hear @WalterBright 's words echoing in my head. "This needs a DIP \*leaves conversation forever\*".

**Comment by tgehr** [comment]
@look-at-me That's nonsense. The DIP specified ho

> > If we are fine with DMD's current behavior, we don't need a patch and the bug report should be closed as invalid.
> 
> The feature shouldn't be using a lambda in the first place.

Nonsense.

> The reason it does is because the person implementing it took a shortcut.
> ...

The `static foreach` DIP explicitly specifies that CTFE is used to create an array for aggregate types other than arrays. This is not a shortcut, this is the best way to ensure that `static foreach` behaves just like `foreach`. DMD uses lowering all over the place and Andrei considers it a best practice for specifying and implementing new language features. It's really not my fault that there are bugs in the `@nogc` implementation. Calling an allocating lambda during CTFE should work in any function, `@nogc` or not.

> > Recursively applying attributes to nested functions is a stupid idea as their attributes are inferred anyway.
> 
> I agree, but it's a big-ish breaking change. I can already hear @WalterBright 's words echoing in my head. "This needs a DIP *leaves conversation forever*".

It's a bug, and pragmatically fixing it at most breaks `__traits(compiles, { })` checks, which are broken anyway for attributes (such as `@safe`) that were actually implemented correctly. Just use `__traits(compiles,{ }())`.

**Comment by look-at-me** [comment]
> Nonsense.

What about the following code *screams* at you, this MUST use a lambda?

```D
static foreach(i; 0 .. 1) {}
```

Nothing does, it's an implementation detail that the user should never even notice is there. The only nonsense thing about it is that you don't think it should be fixed.

> The static foreach DIP explicitly specifies that CTFE is used to create an array for aggregate types other than arrays.  This is not a shortcut, this is the best way to ensure that static foreach behaves just like foreach. 

That's the reason why it is so inefficient and uses so much memory. It is a shortcut and it comes at the cost of efficiency and memory. Whatever the pros and cons are doesn't make it not a shortcut.

> Calling an allocating lambda during CTFE should work in any function, @nogc or not.

The problem is a function annotated with @nogc calling the GC. It really doesn't have anything to do with CTFE. Defining a @nogc function and calling the GC from it should cause an error, is the expected behavior. You inadvertently fixed it by the bug you introduced changing DMD's behavior with scoped attributes.

**Comment by tgehr** [comment]
> > Nonsense.
> 
> What about the following code _screams_ at you, this MUST use a lambda?
> 
> ```d
> static foreach(i; 0 .. 1) {}
> ```
> 
> Nothing does, it's an implementation detail that the user should never even notice is there.

Agreed. It does not _have_ to use a lambda.

> The only nonsense thing about it is that you don't think it should be fixed.
> ...

Don't tell me what I think, especially when being mistaken. Thanks. You are the one who decided reverting the hotfix instead of documenting it was appropriate.

> > The static foreach DIP explicitly specifies that CTFE is used to create an array for aggregate types other than arrays.  This is not a shortcut, this is the best way to ensure that static foreach behaves just like foreach.
> 
> That's the reason why it is so inefficient and uses so much memory.

No. CTFE is not _that_ slow (and it shouldn't be as slow and memory inefficient as it is now in the first place). Presumably, the reason is that big tuple expressions are inefficient. This should be reasonably easy to improve, by modifying the tuple `foreach` code such that it can accept an array instead of a tuple expression, if we do decide that it is in fact okay for tuple expressions to be so slow.

> It is a shortcut and it comes at the cost of efficiency and memory.

It's just not optimized prematurely. This is like saying implementing DMD in D instead of assembly is a shortcut that comes at the cost of efficiency and memory. It's not a useful statement.

> Whatever the pros and cons are doesn't make it not a shortcut.
> ...

The initial implementation has to be obviously correct. Also, if new kinds of aggregates are ever added to `foreach`, `static foreach` gets them for free. Where is the actual, reasonable code that uses `static foreach` on huge ranges?

> > Calling an allocating lambda during CTFE should work in any function, @nogc or not.
> 
> The problem is a function annotated with @nogc calling the GC.

There is no such thing happening, no matter what the current (buggy) DMD has to say about it.

> It really doesn't have anything to do with CTFE.

CTFE is a sufficient but not a necessary condition for it to work.

> Defining a @nogc function and calling the GC from it should cause an error, is the expected behavior. You inadvertently fixed it by the bug you introduced

There is no `@nogc`-related bug in the `static foreach` implementation.

> changing DMD's behavior with scoped attributes.

No, I _fixed_ it only for `@nogc`. `@safe` already works that way (i.e., correctly).

I really don't understand what you are getting out of wasting my time like this.

**Comment by look-at-me** [comment]
> > The only nonsense thing about it is that you don't think it should be fixed.
> > ...
> 
> Don't tell me what I think, especially when being mistaken. Thanks. You are the one who decided reverting the hotfix instead of documenting it was appropriate.

I didn't tell you what you think. You said it yourself, or did you forget what you said?

> If we are fine with DMD's current behavior, we don't need a patch and **the bug report should be closed as invalid**. 
https://github.com/dlang/dmd/pull/9922#issuecomment-510129798

> > > The static foreach DIP explicitly specifies that CTFE is used to create an array for aggregate types other than arrays.  This is not a shortcut, this is the best way to ensure that static foreach behaves just like foreach.
> > 
> > 
> > That's the reason why it is so inefficient and uses so much memory.
> 
> No. CTFE is not _that_ slow (and it shouldn't be as slow and memory inefficient as it is now in the first place). Presumably, the reason is that big tuple expressions are inefficient. This should be reasonably easy to improve, by modifying the tuple `foreach` code such that it can accept an array instead of a tuple expression, if we do decide that it is in fact okay for tuple expressions to be so slow.

It easily takes hours to compile and uses 10's of GB of RAM with even just an empty `static foreach` statement for a few thousand iterations.

> > It is a shortcut and it comes at the cost of efficiency and memory.
> 
> It's just not optimized prematurely. This is like saying implementing DMD in D instead of assembly is a shortcut that comes at the cost of efficiency and memory. It's not a useful statement.

Honestly, it feels like you are just arguing against the definition of shortcut. Do you not agree with the definition of shortcut, that it is a short alternative route, ei takes less time?

> > > Calling an allocating lambda during CTFE should work in any function, @nogc or not.
> > 
> > 
> > The problem is a function annotated with @nogc calling the GC.
> 
> There is no such thing happening, no matter what the current (buggy) DMD has to say about it.

The lambda  the shortcut implementation is annotated with @nogc. It also uses the GC. This is where the core problem of the issue lies.

> > Defining a @nogc function and calling the GC from it should cause an error, is the expected behavior. You inadvertently fixed it by the bug you introduced
> 
> There is no `@nogc`-related bug in the `static foreach` implementation.

Yes there is, here you go: https://issues.dlang.org/show_bug.cgi?id=18439

**Comment by tgehr** [comment]
> > > The only nonsense thing about it is that you don't think it should be fixed.
> > > ...
> > 
> > 
> > Don't tell me what I think, especially when being mistaken. Thanks. You are the one who decided reverting the hotfix instead of documenting it was appropriate.
> 
> I didn't tell you what you think. You said it yourself, or did you forget what you said?
> 
> > If we are fine with DMD's current behavior, we don't need a patch and **the bug report should be closed as invalid**.
> > [#9922 (comment)](https://github.com/dlang/dmd/pull/9922#issuecomment-510129798)
> 

Yes. If. But I'm not fine with it. Neither are you. I was just pointing out that complaining that a bug fix changes the compiler behavior makes no sense. It's called reductio ad absurdum and is a common technique in logic.

**Comment by tgehr** [comment]
> ...
> 
> It easily takes hours to compile and uses 10's of GB of RAM with even just an empty `static foreach` statement for a few thousand iterations.
> 
`static foreach(i;0..5000){}`
This finishes in less than 2 seconds on my machine.

The running time scales quadratically, but this is the case even if you just compute `iota(n).array` in CTFE. This should be fixed.

**Pull request merged**